File: src/other-languages/java/alap/ValidateConfig.java
// Copyright 2026 Daniel Smith
// Licensed under the Apache License, Version 2.0
// See https://www.apache.org/licenses/LICENSE-2.0

package alap;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.IdentityHashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;
import java.util.logging.Logger;
import java.util.regex.Pattern;

/**
 * Config validation — Java port of src/core/validateConfig.ts.
 *
 * <p>Takes an untrusted config map and returns a deeply-immutable,
 * provenance-stamped copy. Mirrors the 3.2 reference behaviour:
 *
 * <ul>
 *   <li>deep-clones the input (rejects callables, Java arrays,
 *       custom class instances, cycles, over-bound structures);</li>
 *   <li>rejects callable-valued protocol handlers with
 *       {@link ConfigMigrationError};</li>
 *   <li>stamps each validated link with the caller-supplied provenance tier;</li>
 *   <li>enforces the hooks allowlist against non-author tiers (fail-closed
 *       when {@code settings.hooks} is not declared);</li>
 *   <li>sanitizes every URL-bearing field ({@code url}, {@code image},
 *       {@code thumbnail}, and any {@code meta.*Url} key) through
 *       {@link SanitizeUrl#sanitize(String)};</li>
 *   <li>strips {@code __proto__} / {@code constructor} / {@code prototype}
 *       keys (plus the Python-port dunders retained for cross-port
 *       parity) from all map-shaped fields, including nested
 *       {@code link.meta};</li>
 *   <li>rejects hyphens in link IDs, tag names, macro names, and
 *       searchPattern keys;</li>
 *   <li>deep-freezes the returned config via {@link Collections#unmodifiableMap}
 *       / {@link Collections#unmodifiableList};</li>
 *   <li>short-circuits when re-validating a config it has already
 *       produced (so storage-tier stamps are not overwritten to
 *       {@code "author"}).</li>
 * </ul>
 */
public final class ValidateConfig {

    private static final Logger LOG = Logger.getLogger(ValidateConfig.class.getName());

    public static final Set<String> BLOCKED_KEYS = Set.of(
        // JS prototype-pollution set (parity with TS).
        "__proto__", "constructor", "prototype",
        // Python-specific dunders retained for cross-port parity; harmless
        // in Java but the single shared blocklist means auditors do not
        // need a per-language cheat sheet.
        "__class__", "__bases__", "__mro__", "__subclasses__"
    );

    private static final Pattern URL_KEY_PATTERN =
        Pattern.compile("url$", Pattern.CASE_INSENSITIVE);

    // Idempotence tracker. IdentityHashMap so structurally-equal configs
    // do not collide; synchronized for multi-threaded safety. The entries
    // leak for the lifetime of the process, bounded by the count of
    // unique configs validated.
    private static final Map<Object, Boolean> VALIDATED =
        Collections.synchronizedMap(new IdentityHashMap<>());

    private ValidateConfig() {}

    /**
     * Reject callable-valued protocol handlers in {@code config}.
     *
     * <p>Handlers must be registered via the runtime registry, not
     * embedded in the config. Thrown loudly at the validate front door
     * so the shape mismatch surfaces with the exact field name.
     */
    public static void assertNoHandlersInConfig(Map<String, Object> config) {
        if (config == null) return;
        if (!(config.get("protocols") instanceof Map<?, ?> protocols)) return;

        for (Map.Entry<?, ?> entry : protocols.entrySet()) {
            if (!(entry.getKey() instanceof String name)) continue;
            if (!(entry.getValue() instanceof Map<?, ?> entryMap)) continue;
            for (String field : List.of("generate", "filter", "handler")) {
                Object v = entryMap.get(field);
                if (isCallable(v)) {
                    throw new ConfigMigrationError(
                        "config[\"protocols\"][\"" + name + "\"][\"" + field +
                        "\"] is a callable — handlers must be registered separately " +
                        "via the runtime registry, not embedded in the config. " +
                        "See docs/handlers-out-of-config.md.");
                }
            }
        }
    }

    private static boolean isCallable(Object v) {
        return v instanceof Runnable
            || v instanceof Function
            || v instanceof BiFunction
            || v instanceof Consumer
            || v instanceof Supplier
            || v instanceof Predicate;
    }

    /**
     * Single source of truth for URL-scheme sanitization on a link.
     *
     * <p>Scans {@code url}, {@code image}, {@code thumbnail}, and any
     * {@code meta} key whose name ends with {@code url}
     * (case-insensitive), passing each through
     * {@link SanitizeUrl#sanitize(String)}. Strips {@link #BLOCKED_KEYS}
     * from {@code meta} during the pass.
     */
    public static Map<String, Object> sanitizeLinkUrls(Map<String, Object> link) {
        Map<String, Object> out = new LinkedHashMap<>(link);
        if (link.get("url") instanceof String s) {
            out.put("url", SanitizeUrl.sanitize(s));
        }
        if (link.get("image") instanceof String s) {
            out.put("image", SanitizeUrl.sanitize(s));
        }
        if (link.get("thumbnail") instanceof String s) {
            out.put("thumbnail", SanitizeUrl.sanitize(s));
        }
        if (link.get("meta") instanceof Map<?, ?> metaMap) {
            Map<String, Object> safeMeta = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : metaMap.entrySet()) {
                if (!(e.getKey() instanceof String mk)) continue;
                if (BLOCKED_KEYS.contains(mk)) continue;
                Object mv = e.getValue();
                if (mv instanceof String mvs && URL_KEY_PATTERN.matcher(mk).find()) {
                    safeMeta.put(mk, SanitizeUrl.sanitize(mvs));
                } else {
                    safeMeta.put(mk, mv);
                }
            }
            out.put("meta", safeMeta);
        }
        return out;
    }

    /**
     * Validate and sanitize {@code raw} from an untrusted source, stamping
     * each link with {@code "author"} provenance.
     */
    public static Map<String, Object> validate(Map<String, Object> raw) {
        return validate(raw, "author");
    }

    /**
     * Validate and sanitize {@code raw}, stamping each link with
     * {@code provenance} tier.
     */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> validate(Map<String, Object> raw, String provenance) {
        if (raw == null) {
            throw new IllegalArgumentException("config is null");
        }

        // Idempotence short-circuit.
        if (VALIDATED.containsKey(raw)) {
            return raw;
        }

        // Reject callable protocol handlers before any further processing.
        assertNoHandlersInConfig(raw);

        // Detach + validate shape via DeepClone.
        Map<String, Object> cloned = (Map<String, Object>) DeepClone.call(raw);

        // Hook allowlist pulled from settings up front.
        Set<String> hookAllowlist = null;
        if (cloned.get("settings") instanceof Map<?, ?> settingsRaw
                && settingsRaw.get("hooks") instanceof List<?> hooksList) {
            hookAllowlist = new HashSet<>();
            for (Object h : hooksList) {
                if (h instanceof String hs) hookAllowlist.add(hs);
            }
        }

        // --- allLinks (required) ---
        if (!(cloned.get("allLinks") instanceof Map<?, ?> allLinksMap)) {
            throw new IllegalArgumentException("allLinks must be a non-null object");
        }

        Map<String, Object> sanitizedLinks = new LinkedHashMap<>();

        for (Map.Entry<?, ?> entry : allLinksMap.entrySet()) {
            if (!(entry.getKey() instanceof String key)) continue;

            if (BLOCKED_KEYS.contains(key)) continue;
            if (key.contains("-")) {
                LOG.warning("validate: skipping allLinks[\"" + key + "\"] — hyphens not allowed in item IDs");
                continue;
            }

            if (!(entry.getValue() instanceof Map<?, ?> link)) {
                LOG.warning("validate: skipping allLinks[\"" + key + "\"] — not a valid link object");
                continue;
            }

            if (!(link.get("url") instanceof String urlStr)) {
                LOG.warning("validate: skipping allLinks[\"" + key + "\"] — missing or invalid url");
                continue;
            }

            // Shape via whitelist.
            Map<String, Object> shaped = new LinkedHashMap<>();
            shaped.put("url", urlStr);
            if (link.get("label") instanceof String s) shaped.put("label", s);
            if (link.get("cssClass") instanceof String s) shaped.put("cssClass", s);
            if (link.get("image") instanceof String s) shaped.put("image", s);
            if (link.get("altText") instanceof String s) shaped.put("altText", s);
            if (link.get("targetWindow") instanceof String s) shaped.put("targetWindow", s);
            if (link.get("description") instanceof String s) shaped.put("description", s);
            if (link.get("thumbnail") instanceof String s) shaped.put("thumbnail", s);

            // Tags — strings only, reject hyphens.
            if (link.get("tags") instanceof List<?> tagsList) {
                List<String> cleanTags = new ArrayList<>();
                for (Object t : tagsList) {
                    if (!(t instanceof String tagStr)) continue;
                    if (tagStr.contains("-")) {
                        LOG.warning("validate: allLinks[\"" + key + "\"] — stripping tag \""
                            + tagStr + "\" (hyphens not allowed)");
                        continue;
                    }
                    cleanTags.add(tagStr);
                }
                shaped.put("tags", cleanTags);
            }

            // Hooks — tier-aware allowlist enforcement.
            if (link.get("hooks") instanceof List<?> hooksList) {
                List<String> stringHooks = new ArrayList<>();
                for (Object h : hooksList) {
                    if (h instanceof String hs) stringHooks.add(hs);
                }
                if ("author".equals(provenance)) {
                    if (!stringHooks.isEmpty()) shaped.put("hooks", stringHooks);
                } else if (hookAllowlist != null) {
                    List<String> allowed = new ArrayList<>();
                    for (String h : stringHooks) {
                        if (hookAllowlist.contains(h)) {
                            allowed.add(h);
                        } else {
                            LOG.warning("validate: allLinks[\"" + key + "\"] — stripping hook \""
                                + h + "\" not in settings.hooks allowlist (tier: " + provenance + ")");
                        }
                    }
                    if (!allowed.isEmpty()) shaped.put("hooks", allowed);
                } else if (!stringHooks.isEmpty()) {
                    LOG.warning("validate: allLinks[\"" + key + "\"] — dropping "
                        + stringHooks.size() + " hook(s) on " + provenance
                        + "-tier link; declare settings.hooks to allow specific keys");
                }
            }

            if (link.get("guid") instanceof String s) shaped.put("guid", s);
            if (link.containsKey("createdAt")) shaped.put("createdAt", link.get("createdAt"));

            // Meta — copy with nested BLOCKED_KEYS filter. (sanitizeLinkUrls
            // runs a second pass that also strips blocked keys and
            // sanitises *Url fields; this first pass makes sure
            // shaped["meta"] is already a fresh map.)
            if (link.get("meta") instanceof Map<?, ?> metaMap) {
                Map<String, Object> safeMeta = new LinkedHashMap<>();
                for (Map.Entry<?, ?> me : metaMap.entrySet()) {
                    if (!(me.getKey() instanceof String mk)) continue;
                    if (BLOCKED_KEYS.contains(mk)) continue;
                    safeMeta.put(mk, me.getValue());
                }
                shaped.put("meta", safeMeta);
            }

            // Single source of truth for URL-field sanitization.
            Map<String, Object> finalLink = sanitizeLinkUrls(shaped);

            // Stamp provenance AFTER the whitelist pass — since shaped
            // was built from a fixed set of known keys, an incoming
            // config cannot pre-stamp itself via a forged _provenance field.
            LinkProvenance.stamp(finalLink, provenance);

            sanitizedLinks.put(key, finalLink);
        }

        // --- settings (optional) ---
        Map<String, Object> settings = filterBlocked(cloned.get("settings"));

        // --- macros (optional) ---
        Map<String, Object> macros = null;
        if (cloned.get("macros") instanceof Map<?, ?> macrosRaw) {
            macros = new LinkedHashMap<>();
            for (Map.Entry<?, ?> entry : macrosRaw.entrySet()) {
                if (!(entry.getKey() instanceof String name)) continue;
                if (BLOCKED_KEYS.contains(name)) continue;
                if (name.contains("-")) {
                    LOG.warning("validate: skipping macro \"" + name + "\" — hyphens not allowed");
                    continue;
                }
                if (entry.getValue() instanceof Map<?, ?> macroMap
                        && macroMap.get("linkItems") instanceof String) {
                    macros.put(name, new LinkedHashMap<>(macroMap));
                }
            }
        }

        // --- searchPatterns (optional) ---
        Map<String, Object> searchPatterns = null;
        if (cloned.get("searchPatterns") instanceof Map<?, ?> spRaw) {
            searchPatterns = new LinkedHashMap<>();
            for (Map.Entry<?, ?> entry : spRaw.entrySet()) {
                if (!(entry.getKey() instanceof String spKey)) continue;
                if (BLOCKED_KEYS.contains(spKey)) continue;
                if (spKey.contains("-")) {
                    LOG.warning("validate: skipping searchPattern \"" + spKey + "\" — hyphens not allowed");
                    continue;
                }

                if (entry.getValue() instanceof String patStr) {
                    Config.RegexValidation v = ValidateRegex.validate(patStr);
                    if (v.safe()) {
                        searchPatterns.put(spKey, patStr);
                    } else {
                        LOG.warning("validate: removing searchPattern \"" + spKey + "\" — " + v.reason());
                    }
                } else if (entry.getValue() instanceof Map<?, ?> patMap
                        && patMap.get("pattern") instanceof String pat) {
                    Config.RegexValidation v = ValidateRegex.validate(pat);
                    if (v.safe()) {
                        searchPatterns.put(spKey, new LinkedHashMap<>(patMap));
                    } else {
                        LOG.warning("validate: removing searchPattern \"" + spKey + "\" — " + v.reason());
                    }
                }
            }
        }

        // --- protocols (optional, data-only since 3.2) ---
        Map<String, Object> protocols = null;
        if (cloned.get("protocols") instanceof Map<?, ?> protoRaw) {
            protocols = new LinkedHashMap<>();
            for (Map.Entry<?, ?> entry : protoRaw.entrySet()) {
                if (!(entry.getKey() instanceof String pkey)) continue;
                if (BLOCKED_KEYS.contains(pkey)) continue;
                protocols.put(pkey, entry.getValue());
            }
        }

        // Assemble.
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("allLinks", sanitizedLinks);
        if (settings != null && !settings.isEmpty()) result.put("settings", settings);
        if (macros != null && !macros.isEmpty()) result.put("macros", macros);
        if (searchPatterns != null && !searchPatterns.isEmpty()) result.put("searchPatterns", searchPatterns);
        if (protocols != null && !protocols.isEmpty()) result.put("protocols", protocols);

        // Deep-freeze children and the top level.
        Map<String, Object> frozen = (Map<String, Object>) deepFreeze(result);
        VALIDATED.put(frozen, Boolean.TRUE);
        return frozen;
    }

    /**
     * Recursively wrap maps and lists in unmodifiable views. Strings,
     * numbers, booleans, and nulls pass through unchanged.
     */
    @SuppressWarnings("unchecked")
    private static Object deepFreeze(Object obj) {
        if (obj instanceof Map<?, ?> map) {
            Map<String, Object> out = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (e.getKey() instanceof String k) {
                    out.put(k, deepFreeze(e.getValue()));
                }
            }
            return Collections.unmodifiableMap(out);
        }
        if (obj instanceof List<?> list) {
            List<Object> out = new ArrayList<>(list.size());
            for (Object item : list) out.add(deepFreeze(item));
            return Collections.unmodifiableList(out);
        }
        return obj;
    }

    private static Map<String, Object> filterBlocked(Object raw) {
        if (!(raw instanceof Map<?, ?> map)) return null;
        Map<String, Object> result = new LinkedHashMap<>();
        for (Map.Entry<?, ?> entry : map.entrySet()) {
            if (entry.getKey() instanceof String key && !BLOCKED_KEYS.contains(key)) {
                result.put(key, entry.getValue());
            }
        }
        return result;
    }
}

File: src/other-languages/ruby/lib/alap/validate_config.rb
# frozen_string_literal: true

# Copyright 2026 Daniel Smith
# Licensed under the Apache License, Version 2.0
# See https://www.apache.org/licenses/LICENSE-2.0

require "set"

module Alap
  # Raised when a config has a legacy shape requiring migration.
  #
  # Currently thrown by +Alap::ValidateConfig.assert_no_handlers_in_config+
  # when a +config["protocols"][<name>]["generate" | "filter" | "handler"]+
  # slot holds a callable. Handlers must be registered separately via the
  # runtime registry; the config itself is pure data.
  class ConfigMigrationError < StandardError; end

  # Config validation — Ruby port of src/core/validateConfig.ts.
  #
  # Takes an untrusted config hash and returns a frozen,
  # provenance-stamped copy. Mirrors the 3.2 reference behaviour:
  #
  # - deep-clones the input (rejects Procs, class instances, Hash/Array
  #   subclasses, cycles);
  # - rejects callable protocol handlers with +ConfigMigrationError+;
  # - stamps each validated link with the caller-supplied provenance tier;
  # - enforces the hooks allowlist against non-author tiers (fail-closed
  #   when +settings.hooks+ is not declared);
  # - sanitizes every URL-bearing field (+url+, +image+, +thumbnail+,
  #   and any +meta.*Url+ key) through +Alap::SanitizeUrl.call+;
  # - strips +__proto__+, +constructor+, +prototype+ keys from all
  #   Hash-shaped fields, including nested +link["meta"]+;
  # - rejects hyphens in link IDs, tag names, macro names, and
  #   searchPattern keys (+-+ is the WITHOUT operator in expressions);
  # - deep-freezes the returned config so handlers see the shape the
  #   validator approved;
  # - short-circuits when re-validating a config it has already produced
  #   (so storage-tier stamps are not overwritten to +"author"+).
  module ValidateConfig
    LINK_FIELD_WHITELIST = Set.new(%w[
      url label tags cssClass image altText targetWindow
      description thumbnail hooks guid createdAt
    ]).freeze

    BLOCKED_KEYS = Set.new(%w[
      __proto__ constructor prototype
      __class__ __bases__ __mro__ __subclasses__
    ]).freeze

    URL_KEY_RE = /url\z/i

    # Module-level WeakMap tracking configs this validator has produced.
    # Weak-keyed so entries auto-clear as the config hashes are GC'd.
    # The WeakMap is the idempotence witness: membership is proof that
    # +call+ produced the hash, so an external caller cannot forge
    # idempotence from outside.
    VALIDATED = ObjectSpace::WeakMap.new

    # Reject callable-valued protocol handlers in +config+.
    #
    # Handlers must be registered via the runtime registry, not embedded
    # in the config. Thrown loudly at the validate front door so the
    # shape mismatch surfaces with the exact field name, not as a
    # missing handler at first dispatch.
    def self.assert_no_handlers_in_config(config)
      return unless config.is_a?(Hash)
      protocols = config["protocols"]
      return unless protocols.is_a?(Hash)

      protocols.each do |name, entry|
        next unless entry.is_a?(Hash)
        %w[generate filter handler].each do |field|
          value = entry[field]
          if value.is_a?(Proc) || value.is_a?(Method) || value.is_a?(UnboundMethod)
            raise ConfigMigrationError,
                  "config[\"protocols\"][#{name.inspect}][#{field.inspect}] is a " \
                  "callable — handlers must be registered separately via the " \
                  "runtime registry, not embedded in the config. " \
                  "See docs/handlers-out-of-config.md."
          end
        end
      end
    end

    # Single source of truth for URL-scheme sanitization on a link.
    #
    # Scans +url+, +image+, +thumbnail+, and any +meta+ key whose name
    # ends with +url+ (case-insensitive), passing each through
    # +Alap::SanitizeUrl.call+. Strips +__proto__+, +constructor+,
    # +prototype+ (plus the Python-port dunders retained for cross-port
    # parity) from +meta+ during the pass.
    def self.sanitize_link_urls(link)
      out = link.dup
      out["url"] = Alap::SanitizeUrl.call(link["url"]) if link["url"].is_a?(String)
      out["image"] = Alap::SanitizeUrl.call(link["image"]) if link["image"].is_a?(String)
      out["thumbnail"] = Alap::SanitizeUrl.call(link["thumbnail"]) if link["thumbnail"].is_a?(String)
      raw_meta = link["meta"]
      if raw_meta.is_a?(Hash)
        safe_meta = {}
        raw_meta.each do |mk, mv|
          next if BLOCKED_KEYS.include?(mk)
          safe_meta[mk] = if mv.is_a?(String) && URL_KEY_RE.match?(mk)
                           Alap::SanitizeUrl.call(mv)
                         else
                           mv
                         end
        end
        out["meta"] = safe_meta
      end
      out
    end

    # Validate and sanitize +config+ from an untrusted source.
    #
    # Returns a frozen copy with each link stamped with +provenance+.
    # Raises +ArgumentError+ on structural invalidity,
    # +ConfigMigrationError+ on callable handlers, or
    # +Alap::DeepClone::Error+ on non-data types / cycles / over-bound
    # structures.
    def self.call(config, provenance: "author")
      # Idempotence short-circuit: a pre-validated config has its
      # original provenance stamps (including storage-tier stamps from
      # storage adapters); re-running the pipeline would overwrite them
      # to "author". VALIDATED membership is proof we produced this
      # object, so the caller cannot forge idempotence from outside.
      return config if config.is_a?(Hash) && VALIDATED[config]

      raise ArgumentError, "Invalid config: expected a Hash" unless config.is_a?(Hash)

      # Reject callable protocol handlers before any further processing
      # so the migration error surfaces at the exact field name, not as
      # a generic "not data" from DeepClone.
      assert_no_handlers_in_config(config)

      # Detach from caller; DeepClone rejects callables / class
      # instances / cycles / non-String keys / over-bound structures.
      raw = Alap::DeepClone.call(config)

      # Hook allowlist pulled from settings up front so the per-link
      # pass below can filter non-author-tier hooks against it.
      raw_settings = raw["settings"]
      hook_allowlist =
        if raw_settings.is_a?(Hash) && raw_settings["hooks"].is_a?(Array)
          Set.new(raw_settings["hooks"].select { |h| h.is_a?(String) })
        end

      # --- allLinks (required) ---------------------------------------
      raw_links = raw["allLinks"]
      raise ArgumentError, "Invalid config: allLinks must be a non-null Hash" unless raw_links.is_a?(Hash)

      sanitized_links = {}
      raw_links.each do |key, link|
        next if BLOCKED_KEYS.include?(key)

        if key.include?("-")
          warn "[Alap] validate_config: skipping allLinks[#{key.inspect}] — " \
               "hyphens are not allowed in item IDs. Use underscores " \
               "instead. The \"-\" character is the WITHOUT operator in expressions."
          next
        end

        unless link.is_a?(Hash)
          warn "[Alap] validate_config: skipping allLinks[#{key.inspect}] — not a valid link object"
          next
        end

        unless link["url"].is_a?(String)
          warn "[Alap] validate_config: skipping allLinks[#{key.inspect}] — missing or invalid url"
          next
        end

        # Tags — strings only, reject hyphens.
        tags = nil
        if link.key?("tags")
          if link["tags"].is_a?(Array)
            tags = link["tags"].select do |t|
              next false unless t.is_a?(String)

              if t.include?("-")
                warn "[Alap] validate_config: allLinks[#{key.inspect}] — " \
                     "stripping tag #{t.inspect} (hyphens not allowed in tags). " \
                     "Use underscores instead."
                next false
              end

              true
            end
          else
            warn "[Alap] validate_config: allLinks[#{key.inspect}].tags is not an Array — ignoring"
          end
        end

        # Shape via whitelist.
        shaped = { "url" => link["url"] }
        shaped["label"] = link["label"] if link["label"].is_a?(String)
        shaped["tags"] = tags unless tags.nil?
        shaped["cssClass"] = link["cssClass"] if link["cssClass"].is_a?(String)
        shaped["image"] = link["image"] if link["image"].is_a?(String)
        shaped["altText"] = link["altText"] if link["altText"].is_a?(String)
        shaped["targetWindow"] = link["targetWindow"] if link["targetWindow"].is_a?(String)
        shaped["description"] = link["description"] if link["description"].is_a?(String)
        shaped["thumbnail"] = link["thumbnail"] if link["thumbnail"].is_a?(String)

        # Hooks — tier-aware allowlist enforcement.
        if link["hooks"].is_a?(Array)
          string_hooks = link["hooks"].select { |h| h.is_a?(String) }
          if provenance == "author"
            shaped["hooks"] = string_hooks unless string_hooks.empty?
          elsif hook_allowlist
            allowed = []
            string_hooks.each do |h|
              if hook_allowlist.include?(h)
                allowed << h
              else
                warn "[Alap] validate_config: allLinks[#{key.inspect}] — " \
                     "stripping hook #{h.inspect} not in settings.hooks " \
                     "allowlist (tier: #{provenance})"
              end
            end
            shaped["hooks"] = allowed unless allowed.empty?
          elsif !string_hooks.empty?
            warn "[Alap] validate_config: allLinks[#{key.inspect}] — " \
                 "dropping #{string_hooks.length} hook(s) on #{provenance}-tier " \
                 "link; declare settings.hooks to allow specific keys"
          end
        end

        shaped["guid"] = link["guid"] if link["guid"].is_a?(String)
        shaped["createdAt"] = link["createdAt"] if link.key?("createdAt")

        # Meta — copy with nested BLOCKED_KEYS filter. (sanitize_link_urls
        # will run a second pass that also strips blocked keys and
        # sanitises *Url fields; this first pass makes sure shaped["meta"]
        # is already a fresh hash.)
        if link["meta"].is_a?(Hash)
          raw_meta = link["meta"]
          safe_meta = {}
          raw_meta.each do |mk, mv|
            next if BLOCKED_KEYS.include?(mk)
            safe_meta[mk] = mv
          end
          shaped["meta"] = safe_meta
        end

        # Single source of truth for URL-field sanitization.
        final_link = sanitize_link_urls(shaped)

        # Stamp provenance AFTER the whitelist pass — since shaped was
        # built from a fixed set of known keys, an incoming config
        # cannot pre-stamp itself via a forged _provenance field.
        Alap::LinkProvenance.stamp(final_link, provenance)

        sanitized_links[key] = final_link
      end

      # --- settings (optional) ---------------------------------------
      settings = nil
      if raw_settings.is_a?(Hash)
        settings = {}
        raw_settings.each do |skey, sval|
          next if BLOCKED_KEYS.include?(skey)
          settings[skey] = sval
        end
      end

      # --- macros (optional) -----------------------------------------
      macros = nil
      raw_macros = raw["macros"]
      if raw_macros.is_a?(Hash)
        macros = {}
        raw_macros.each do |mkey, macro|
          next if BLOCKED_KEYS.include?(mkey)

          if mkey.include?("-")
            warn "[Alap] validate_config: skipping macro #{mkey.inspect} — " \
                 "hyphens are not allowed in macro names. Use underscores " \
                 "instead. The \"-\" character is the WITHOUT operator in expressions."
            next
          end

          if macro.is_a?(Hash) && macro["linkItems"].is_a?(String)
            macros[mkey] = macro
          else
            warn "[Alap] validate_config: skipping macro #{mkey.inspect} — invalid shape"
          end
        end
      end

      # --- searchPatterns (optional) --------------------------------
      search_patterns = nil
      raw_patterns = raw["searchPatterns"]
      if raw_patterns.is_a?(Hash)
        search_patterns = {}
        raw_patterns.each do |pkey, entry|
          next if BLOCKED_KEYS.include?(pkey)

          if pkey.include?("-")
            warn "[Alap] validate_config: skipping searchPattern #{pkey.inspect} — " \
                 "hyphens are not allowed in pattern keys. Use underscores " \
                 "instead. The \"-\" character is the WITHOUT operator in expressions."
            next
          end

          if entry.is_a?(Hash) && entry["pattern"].is_a?(String)
            validation = Alap::ValidateRegex.call(entry["pattern"])
            if validation["safe"]
              search_patterns[pkey] = entry
            else
              warn "[Alap] validate_config: removing searchPattern " \
                   "#{pkey.inspect} — #{validation["reason"]}"
            end
            next
          end

          if entry.is_a?(String)
            validation = Alap::ValidateRegex.call(entry)
            if validation["safe"]
              search_patterns[pkey] = entry
            else
              warn "[Alap] validate_config: removing searchPattern " \
                   "#{pkey.inspect} — #{validation["reason"]}"
            end
            next
          end

          warn "[Alap] validate_config: skipping searchPattern #{pkey.inspect} — invalid shape"
        end
      end

      # --- protocols (optional, data-only since 3.2) ---------------
      protocols = nil
      raw_protocols = raw["protocols"]
      if raw_protocols.is_a?(Hash)
        protocols = {}
        raw_protocols.each do |pkey, pval|
          next if BLOCKED_KEYS.include?(pkey)
          protocols[pkey] = pval
        end
      end

      # Assemble, freeze, track.
      result = { "allLinks" => sanitized_links }
      result["settings"] = settings if settings
      result["macros"] = macros if macros
      result["searchPatterns"] = search_patterns if search_patterns
      result["protocols"] = protocols if protocols

      Alap::DeepFreeze.call(result)
      VALIDATED[result] = true
      result
    end
  end
end
